modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/DropdownMenu.js

```diff
@@ -0,0 +1,42 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import ClayButton from '@clayui/button';
+import {ClayDropDownWithItems} from '@clayui/drop-down';
+import ClayIcon from '@clayui/icon';
+import React from 'react';
+
+export default function NavigationBar({
+	componentId,
+	cssClass,
+	icon,
+	items,
+	label,
+	locale,
+	portletId,
+	portletNamespace,
+	...otherProps
```

Comment:
This is an interesting gotcha...

Since we pass `...otherProps` to `<ClayButton>`, we want to strip out the props that don't belong to `<ClayButton>` or `<button>`. Otherwise, we eventually generate something like:

```html
<button componentid="foo" locale="[Object object]" portletId="_com_liferay..." />
```

Not sure if there's a better way to do this... 🤔

Reply:
@wincent, thoughts? `eslint` will complain here if we leave those params unused, so not sure what's the best pattern to strip out some props and prevent them from going to `<button>`

Reply:
This is similar but not identical to [this question](https://github.com/liferay/liferay-frontend-guidelines/issues/79) about stripping some properties from an object and keeping a reference to the rest. There's the syntactic question of what is the nicest way to juggle/extract props, and the semantic one of how we determine *which* props matter.

Just to make this concrete, let's say that `ClayButton` accepts these props:

- `alert?: boolean`
- `borderless?: boolean`
- `block?: boolean`
- `displayType?: 'primary' | 'secondary' | 'link' | 'unstyled'`
- `monospaced?: boolean`
- `outline?: boolean`
- `small?: boolean`

In addition to [most of these](https://github.com/microsoft/TypeScript/blob/a60feb2d3983e4b57a257acfefcc8477e66b9425/src/lib/dom.generated.d.ts#L6304) ([etc](https://github.com/microsoft/TypeScript/blob/a60feb2d3983e4b57a257acfefcc8477e66b9425/src/lib/dom.generated.d.ts#L6566)) `button` ones:

- `disabled: boolean`
- `name: string`
- `title: string`
- `dir: string`
- ... etc

I don't think there's any *cheap* way to magically filter the props. If we're worried about this kind of thing:

> Warning: React does not recognize the `componentId` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `componentid` instead. If you accidentally passed it from a parent component, remove it from the DOM element.

... then my inclination would be to start with a whitelist of props that we explicitly forward, and we would drop the rest. If we have to do this in more than one place, the we might want to make an `extractButtonProps()` utility that would be used like this:

```js
<ClayButton className={cssClass} {...extractButtonProps(otherProps)}>
```

(Another option would be to push that down into Clay, but this does feel somewhat like a DXP-specific problem (because of the way we are passing stuff from Java to JS), so it feels like the solution should probably be in DXP.)

Alternatively, we don't do any dynamic extraction: we literally hardcode the list of attributes in the taglib itself. In the end, it's still a whitelist, but it would live in a different place. While maintaining a whitelist sounds onerous, it is effectively what React is doing too, which is why it can show messages like the warning above — it knows what props are valid DOM element props.

And yet another option: we could leverage the TypeScript DOM types that I already linked to above to autogenerate validator functions, or extractor functions.

Reply:
Hey @wincent, I've just left here `componentId`, `portletId`, and `portletNamespace` to avoid the noise. Being this the first instance of this issue it feels like the cheapest way to get this out and then see what we need and where...

I'm not sure about this in any case, so I'm fine with any modification you want to make to this:
- Remove the params and let React complain for now
- Create a util to extract allowed props here or in Clay

> Alternatively, we don't do any dynamic extraction: we literally hardcode the list of attributes in the taglib itself

I think this would be more challenging than whatever solution we have at a JS level